test/wallet-test.js

```diff
@@ -2279,4 +2299,181 @@ describe('Wallet', function() {
+      }
+    });
+
+    it('should send an update for cloudflare', async () => {
+      const pre = await wallet.getBalance();
+      assert.equal(pre.tx, 1);
+      assert.equal(pre.coin, 1);
+      assert.equal(pre.unconfirmed, lockup);
+      assert.equal(pre.confirmed, lockup);
+      assert.equal(pre.ulocked, lockup);
+      assert.equal(pre.clocked, lockup);
+
+      const records = Resource.fromJSON({
+        records: [{type: 'NS', ns: 'ns1.easyhandshake.com.'}]
+      });
+
+      update = await wallet.sendUpdate('cloudflare', records);
+      const entry = nextEntry(wdb);
+      await wdb.addBlock(entry, [update]);
+
+      const ns = await wallet.getNameState(nameHash);
+      const json = ns.getJSON(wdb.state.height, network);
+      assert.equal(json.name, 'cloudflare');
+
+      const resource = Resource.decode(ns.data);
+      assert.deepEqual(records.toJSON(), resource.toJSON());
+
+      // The unconfirmed and confirmed values should
+      // take into account the transaction fee. Assert
+      // against the value of the newly created output.
+      const val = update.output(1).value;
+      const post = await wallet.getBalance();
+      assert.equal(post.tx, 2);
+      assert.equal(post.coin, 2);
+      assert.equal(post.unconfirmed, val);
+      assert.equal(post.confirmed, val);
+      assert.equal(post.ulocked, 0);
+      assert.equal(post.clocked, 0);
+    });
+
+    it('should remove a block and update balances correctly', async () => {
+      const val = update.output(1).value;
+      const pre = await wallet.getBalance();
+      assert.equal(pre.tx, 2);
+      assert.equal(pre.coin, 2);
+      assert.equal(pre.unconfirmed, val);
+      assert.equal(pre.confirmed, val);
+      assert.equal(pre.ulocked, 0);
+      assert.equal(pre.clocked, 0);
+
+      const cur = curEntry(wdb);
+      await wdb.removeBlock(cur);
+
+      const post = await wallet.getBalance();
+      assert.equal(post.tx, 2);
+      assert.equal(post.coin, 2);
+      // The unconfirmed balance includes value in the mempool
+      // and the chain itself. The reorg'd tx can be included
+      // in another block so the unconfirmed total does not
+      // include the tx fee. That value has been effectively
+      // spent already.
+      assert.equal(post.unconfirmed, val);
+      assert.equal(post.confirmed, lockup);
+      assert.equal(post.ulocked, 0);
+      assert.equal(post.clocked, lockup);
+    });
+
+    it('should update balances correctly after abandon', async () => {
+      const val = update.output(1).value;
+      const pre = await wallet.getBalance();
+      assert.equal(pre.tx, 2);
+      assert.equal(pre.coin, 2);
+      assert.equal(pre.unconfirmed, val);
+      assert.equal(pre.confirmed, lockup);
+      assert.equal(pre.ulocked, 0);
+      assert.equal(pre.clocked, lockup);
+
+      // Adding without a block will mark the transaction
+      // as pending, meaning that it was seen in the mempool.
+      await wallet.txdb.add(update);
```

Comment:
This `txdb.add()` is unnecessary. Test passes without it. And that is kinda the point: you assert the tx is in the wdb in the `getBalance` just above, then abandon, then confirm. If you wanna be extra explicit (optional) you can probably test for `txdb.hasTX(hash)` before and after the abandon.